File: algorithms/python/电话号码的字母组合/电话号码的字母组合.py
# Source : https://leetcode.cn/problems/letter-combinations-of-a-phone-number/description/
# Author : Yanan Wang
# Date   : 2023-09-03

class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        n = len(digits)
        if 0 == n:
            return []
        MP = ['', '', 'abc', 'def', 'ghi', 'jkl', 'mno', 'pqrs', 'tuv', 'wxyz']
        ans, path = [], []
        def dfs(index):
            if len(path) == n:
                ans.append(''.join(path))
                return
            for c in MP[int(digits[index])]:
                path.append(c)
                dfs(index + 1)
                path.pop()
        
        dfs(0)
        return ans
        

File: algorithms/cpp/电话号码的字母组合/电话号码的字母组合.cpp
// Source : https://leetcode.cn/problems/letter-combinations-of-a-phone-number/description/
// Author : Yanan Wang
// Date   : 2023-11-29

class Solution {
public:
    string dict[10] = { "", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz" };
    string path;
    vector<string> ans;
    void dfs(string &digits, int level) {
        if (path.size() == digits.size()) {
            ans.push_back(path);
            return;
        }
        string str = dict[digits[level] - '0'];
        for (int i = 0; i < str.size(); ++i) {
            path.push_back(str[i]);
            dfs(digits, level + 1);
            path.pop_back();
        }
    }
    vector<string> letterCombinations(string digits) {
        if (digits.size() == 0) {
            return {};
        }
        dfs(digits, 0);
        return ans;
    }
};